Building this for RPi/OpenELEC:
```
CPP     xbmc/LangInfo.o
CPP     xbmc/MediaSource.o
CPP     xbmc/NfoFile.o
CPP     xbmc/interfaces/python//home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/interfaces/python/generated/AddonModuleXbmc.o
CPP     xbmc/PasswordManager.o
CPP     xbmc/PlayListPlayer.o
CPP     xbmc/PartyModeManager.o
In file included from HTTPImageHandler.cpp:24:0:
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:98:39: error: 'shared_ptr' in namespace 'std' does not name a template type
   static int HandleRequest(const std::shared_ptr<IHTTPRequestHandler>& handler);
                                       ^
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:98:49: error: expected ',' or '...' before '<' token
   static int HandleRequest(const std::shared_ptr<IHTTPRequestHandler>& handler);
                                                 ^
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:99:41: error: 'shared_ptr' in namespace 'std' does not name a template type
   static int FinalizeRequest(const std::shared_ptr<IHTTPRequestHandler>& handler, int responseStatus, struct MHD_Response *response);
                                         ^
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:99:51: error: expected ',' or '...' before '<' token
   static int FinalizeRequest(const std::shared_ptr<IHTTPRequestHandler>& handler, int responseStatus, struct MHD_Response *response);
                                                   ^
In file included from HTTPImageHandler.cpp:24:0:
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:101:54: error: 'shared_ptr' in namespace 'std' does not name a template type
   static int CreateMemoryDownloadResponse(const std::shared_ptr<IHTTPRequestHandler>& handler, struct MHD_Response *&response);
                                                      ^
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:101:64: error: expected ',' or '...' before '<' token
   static int CreateMemoryDownloadResponse(const std::shared_ptr<IHTTPRequestHandler>& handler, struct MHD_Response *&response);
                                                                ^
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:102:60: error: 'shared_ptr' in namespace 'std' does not name a template type
   static int CreateRangedMemoryDownloadResponse(const std::shared_ptr<IHTTPRequestHandler>& handler, struct MHD_Response *&response);
                                                            ^
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:102:70: error: expected ',' or '...' before '<' token
   static int CreateRangedMemoryDownloadResponse(const std::shared_ptr<IHTTPRequestHandler>& handler, struct MHD_Response *&response);
                                                                      ^
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:105:52: error: 'shared_ptr' in namespace 'std' does not name a template type
   static int CreateFileDownloadResponse(const std::shared_ptr<IHTTPRequestHandler>& handler, struct MHD_Response *&response);
                                                    ^
/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/WebServer.h:105:62: error: expected ',' or '...' before '<' token
   static int CreateFileDownloadResponse(const std::shared_ptr<IHTTPRequestHandler>& handler, struct MHD_Response *&response);
                                                              ^
../../../Makefile.include:97: recipe for target 'HTTPImageHandler.o' failed
make[2]: *** [HTTPImageHandler.o] Error 1
make[2]: Leaving directory '/home/neil/projects/OpenELEC.tv/build.OpenELEC-Generic.x86_64-7.0-devel/kodi-17.0-alpha1-e6144a3/xbmc/network/httprequesthandler'
Makefile:415: recipe for target 'xbmc/network/httprequesthandler/httprequesthandlers.a' failed
make[1]: *** [xbmc/network/httprequesthandler/httprequesthandlers.a] Error 2
make[1]: *** Waiting for unfinished jobs....
```

--- a/xbmc/network/WebServer.cpp
+++ b/xbmc/network/WebServer.cpp
@@ -69,6 +69,13 @@ typedef struct ConnectionHandler
   bool isNew;
   IHTTPRequestHandler *requestHandler;
   struct MHD_PostProcessor *postprocessor;
+
+  ConnectionHandler(const std::string& uri)
+    : fullUri(uri)
+    , isNew(true)
+    , postprocessor(nullptr)
+    , requestHandler(nullptr)
+  { }
 } ConnectionHandler;
 
 typedef struct {
@@ -943,18 +950,11 @@ int CWebServer::SendErrorResponse(struct MHD_Connection *connection, int errorTy
 
 void* CWebServer::UriRequestLogger(void *cls, const char *uri)
 {
-  // create a new connection handler
-  ConnectionHandler* conHandler = new ConnectionHandler();
-  conHandler->fullUri = uri;
-  conHandler->isNew = true;
-  conHandler->postprocessor = NULL;
-  conHandler->requestHandler = NULL;
-
   // log the full URI
   CLog::Log(LOGDEBUG, "webserver: request received for %s", uri);
 
-  // return the connection handler so that we can access it in AnswerToConnection as con_cls
-  return conHandler;
+  // create and return a new connection handler
+  return new ConnectionHandler(uri);
 }
 
 #if (MHD_VERSION >= 0x00090200)

--- a/xbmc/network/WebServer.cpp
+++ b/xbmc/network/WebServer.cpp
@@ -69,12 +69,14 @@ typedef struct ConnectionHandler
   bool isNew;
   IHTTPRequestHandler *requestHandler;
   struct MHD_PostProcessor *postprocessor;
+  int errorStatus;
 
   ConnectionHandler(const std::string& uri)
     : fullUri(uri)
     , isNew(true)
     , postprocessor(nullptr)
     , requestHandler(nullptr)
+    , errorStatus(MHD_HTTP_OK)
   { }
 } ConnectionHandler;
 
@@ -361,7 +363,7 @@ int CWebServer::AnswerToConnection(void *cls, struct MHD_Connection *connection,
 
                 delete conHandler->requestHandler;
 
-                return SendErrorResponse(connection, MHD_HTTP_INTERNAL_SERVER_ERROR, methodType);
+                conHandler->errorStatus = MHD_HTTP_INTERNAL_SERVER_ERROR;
               }
             }
           }
@@ -386,18 +388,22 @@ int CWebServer::AnswerToConnection(void *cls, struct MHD_Connection *connection,
       if (conHandler->requestHandler == NULL)
       {
         CLog::Log(LOGERROR, "CWebServer: cannot handle partial HTTP POST for %s request because there is no valid request handler available", url);
-        return SendErrorResponse(connection, MHD_HTTP_INTERNAL_SERVER_ERROR, methodType);
+        conHandler->errorStatus = MHD_HTTP_INTERNAL_SERVER_ERROR;
       }
 
       // we only need to handle POST data if there actually is data left to handle
       if (*upload_data_size > 0)
       {
-        // either use MHD's POST processor
-        if (conHandler->postprocessor != NULL)
-          MHD_post_process(conHandler->postprocessor, upload_data, *upload_data_size);
-        // or simply copy the data to the handler
-        else
-          conHandler->requestHandler->AddPostData(upload_data, *upload_data_size);
+        // if nothing has gone wrong so far, process the given POST data
+        if (conHandler->errorStatus == MHD_HTTP_OK)
+        {
+          // either use MHD's POST processor
+          if (conHandler->postprocessor != NULL)
+            MHD_post_process(conHandler->postprocessor, upload_data, *upload_data_size) == MHD_YES;
+          // or simply copy the data to the handler
+          else
+            conHandler->requestHandler->AddPostData(upload_data, *upload_data_size);
+        }
 
         // signal that we have handled the data
         *upload_data_size = 0;
@@ -413,6 +419,14 @@ int CWebServer::AnswerToConnection(void *cls, struct MHD_Connection *connection,
       {
         if (conHandler->postprocessor != NULL)
           MHD_destroy_post_processor(conHandler->postprocessor);
+
+        // check if something went wrong while handling the POST data
+        if (conHandler->errorStatus != MHD_HTTP_OK)
+        {
+          delete conHandler->requestHandler;
+
+          return SendErrorResponse(connection, conHandler->errorStatus, methodType);
+        }
 
         return HandleRequest(conHandler->requestHandler);
       }

--- a/xbmc/network/httprequesthandler/HTTPJsonRpcHandler.cpp
+++ b/xbmc/network/httprequesthandler/HTTPJsonRpcHandler.cpp
@@ -27,7 +27,7 @@
 #include "utils/log.h"
 #include "utils/Variant.h"
 
-#define MAX_STRING_POST_SIZE 20000
+#define MAX_HTTP_POST_SIZE 65536
 
 bool CHTTPJsonRpcHandler::CanHandleRequest(const HTTPRequest &request)
 {
@@ -127,9 +127,9 @@ bool CHTTPJsonRpcHandler::appendPostData(const char *data, size_t size)
 bool CHTTPJsonRpcHandler::appendPostData(const char *data, unsigned int size)
 #endif
 {
-  if (m_requestData.size() + size > MAX_STRING_POST_SIZE)
+  if (m_requestData.size() + size > MAX_HTTP_POST_SIZE)
   {
-    CLog::Log(LOGERROR, "WebServer: Stopped uploading post since it exceeded size limitations");
+    CLog::Log(LOGERROR, "WebServer: Stopped uploading POST data since it exceeded size limitations (%d)", MAX_HTTP_POST_SIZE);
     return false;
   }
 

